pkg/controller/hypershift/hostedcluster_controller.go

```diff
@@ -0,0 +1,387 @@
+var _ reconcile.Reconciler = &ReconcileHostedcluster{}
+
+// Reconcile the hostedcluster that is in the managed cluster namespace to import the managed cluster.
+//
+// Note: The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileHostedcluster) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request", request)
+	reqLogger.Info("Reconciling HostedCluster")
+	defer reqLogger.Info("Reocileing HostedCluster Done")
+
+	clusterName := request.Name
+	hCluster := &hyperv1.HostedCluster{}
+	err := r.client.Get(ctx, request.NamespacedName, hCluster)
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !hCluster.DeletionTimestamp.IsZero() {
+		// the hostedcluster is deleting, its managed cluster may already be detached (the managed cluster has been deleted,
+		// but the namespace is remained), if it has import finalizer, we remove its namespace
+		return reconcile.Result{}, r.removeImportFinalizer(ctx, hCluster)
+	}
+
+	// TODO: check if UI would do so or not
+	if err := r.ensureManagedClusterCR(ctx, request.NamespacedName); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// add a managed cluster finalizer to the hosted cluster, to handle the managed cluster detach case.
+	if err := r.addClusterImportFinalizer(ctx, hCluster); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// if there is an auto import secret in the managed cluster namespce, we will use the auto import secret to import the cluster
+	_, err = r.kubeClient.CoreV1().Secrets(clusterName).Get(ctx, constants.AutoImportSecretName, metav1.GetOptions{})
+	if err == nil {
+		reqLogger.Info(fmt.Sprintf("The hypershift managed cluster %s has auto import secret, skipped", clusterName))
+		return reconcile.Result{}, nil
+	}
+	if !errors.IsNotFound(err) {
+		return reconcile.Result{}, err
+	}
+
+	if hCluster.Status.KubeConfig == nil {
+		return reconcile.Result{}, fmt.Errorf("hostedcluster's kubeconfig secret is not generated yet")
+	}
+
+	hosteKubeSecertKey := types.NamespacedName{
+		Name:      hCluster.Status.KubeConfig.Name,
+		Namespace: hCluster.Namespace}
+
+	hostedKubeconfigSecret := &corev1.Secret{}
+
+	if err := r.client.Get(ctx, hosteKubeSecertKey, hostedKubeconfigSecret); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	hostedClusterClient, restMapper, err := helpers.GenerateClientFromSecret(hostedKubeconfigSecret)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	importSecretName := fmt.Sprintf("%s-%s", clusterName, constants.ImportSecretNameSuffix)
+	importSecret, err := r.kubeClient.CoreV1().Secrets(clusterName).Get(ctx, importSecretName, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
```

Comment:
I think this should return err to reconcile until the import-secret is created if we do not watch the import-secret.
otherwise there is no event to trigger reconcile since we only watch hostedCluster.

Reply:
for this one, I looked into `helpers.ImportManagedClusterFromSecret` which has checked for the import secret non-exist case:

```
	importSecret *corev1.Secret) error {
	if err := ValidateImportSecret(importSecret); err != nil {
		return err
	}
```

Do you think this is enough or not?

Reply:
it returns nil and cannot go to `helpers.ImportManagedClusterFromSecret` if do not get import secret.